### PR TITLE
[SDXL] Allow SDXL LoRA to be run with less than 16GB of VRAM

### PR DESCRIPTION
**Pull request by patrickvonplaten**
As the issues:
- https://github.com/huggingface/diffusers/issues/4368
- https://github.com/huggingface/diffusers/issues/4377
- https://github.com/huggingface/diffusers/issues/4360

point out training LoRA currently requires a significant amount of memory.
It's quite tough to get SDXL to use less memory given that the unet weights alone are around 6GB in fp16 precision and that we're training on a resolution of 1024 which translates into a latent resolution of 128, things are quite bottlenecked.

In this PR I used the following code snippet:

```py
from diffusers import UNet2DConditionModel
import torch

torch.cuda.set_per_process_memory_fraction(0.4, device="cuda:1")

unet = UNet2DConditionModel.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", subfolder="unet", variant="fp16", torch_dtype=torch.float16)
unet.train()
unet.enable_gradient_checkpointing()
unet = unet.to("cuda:1")

batch_size = 2

sample = torch.randn((1, 4, 128, 128)).half().to(unet.device).repeat(batch_size, 1, 1, 1)
time_ids = (torch.arange(6) / 6)[None, :].half().to(unet.device).repeat(batch_size, 1)
encoder_hidden_states = torch.randn((1, 77, 2048)).half().to(unet.device).repeat(batch_size, 1, 1)
text_embeds = torch.randn((1, 1280)).half().to(unet.device).repeat(batch_size, 1)

out = unet(sample, 1.0, added_cond_kwargs={"time_ids": time_ids, "text_embeds": text_embeds}, encoder_hidden_states=encoder_hidden_states).sample

loss = ((out - sample) ** 2).mean()
loss.backward()

print(torch.cuda.max_memory_allocated(device=unet.device))
```

to optimize gradient checkpointing more by adding it to the Transformer2D model and the cross attention mid block.

Besides gradient checkpointing, using `xformers` instead of `bitsandbytes` for Adam helped quite a bit, so that the final mem requirement for SDXL non-text encoder LoRA training can be brought down to something like 14.8 GB which fits on a T4, thus on a free Google Colab

Screenshot from 2023-08-04 16-49-40

### CODE REVIEW

File: examples/controlnet/train_controlnet_sdxl.py [hunk 1]
```diff
@@ -899,6 +899,7 @@ def load_model_hook(models, input_dir):
 
     if args.gradient_checkpointing:
         controlnet.enable_gradient_checkpointing()
+        unet.enable_gradient_checkpointing()
```
Comment (patrickvonplaten) [inline examples/controlnet/train_controlnet_sdxl.py:902]: some gradients also flow through the unet so it's always safer IMO to enable it as well

File: examples/dreambooth/train_dreambooth_lora.py [hunk 2]
```diff
@@ -839,6 +839,11 @@ def main(args):
         else:
             raise ValueError("xformers is not available. Make sure it is installed correctly")
 
+    if args.gradient_checkpointing:
+        unet.enable_gradient_checkpointing()
```
Comment (patrickvonplaten) [inline examples/dreambooth/train_dreambooth_lora.py:843]: for LoRA gradients still flow through the unet so we should enable gradient checkpointing

File: src/diffusers/models/transformer_2d.py [hunk 3]
```diff
@@ -289,15 +291,28 @@ def forward(
...
-                hidden_states,
-                attention_mask=attention_mask,
-                encoder_hidden_states=encoder_hidden_states,
-                encoder_attention_mask=encoder_attention_mask,
-                timestep=timestep,
-                cross_attention_kwargs=cross_attention_kwargs,
-                class_labels=class_labels,
-            )
+            if self.training and self.gradient_checkpointing:
+                hidden_states = torch.utils.checkpoint.checkpoint(
```
Comment (patrickvonplaten) [inline src/diffusers/models/transformer_2d.py:295]: SDXL has 10 Transformer blocks for the inner resolution blocks so checkpointing those can help quite a bit

File: examples/dreambooth/README_sdxl.md [hunk 4]
```diff
@@ -101,6 +101,24 @@ To better track our training experiments, we're using the following flags in the
 
 Our experiments were conducted on a single 40GB A100 GPU.
 
+### Dog toy example with < 16GB VRAM
+
+By making use of [`gradient_checkpointing`](https://pytorch.org/docs/stable/checkpoint.html) which is natively supported in Diffusers, and the amazing [`xformers`](https://github.com/facebookresearch/xformers) and [`bitsandbytes`](https://github.com/TimDettmers/bitsandbytes) library you can train SDXL LoRAs with less than 16GB of VRAM by adding the following flags to your accelerate launch command:
```
Comment (sayakpaul) [inline examples/dreambooth/README_sdxl.md:106]:
```suggestion
By making use of [`gradient_checkpointing`](https://pytorch.org/docs/stable/checkpoint.html) (which is natively supported in Diffusers), [`xformers`](https://github.com/facebookresearch/xformers), and [`bitsandbytes`](https://github.com/TimDettmers/bitsandbytes) libraries, you can train SDXL LoRAs with less than 16GB of VRAM by adding the following flags to your accelerate launch command:
```

File: examples/dreambooth/README_sdxl.md [hunk 5]
```diff
@@ -101,6 +101,24 @@ To better track our training experiments, we're using the following flags in the
...
+
+```diff
++  --enable_xformers_memory_efficient_attention \
++  --gradient_checkpointing \
++  --use_8bit_adam \
++ --mixed_precision="fp16" \
+```
+
+and making sure that you have the following libraries installed:
```
Comment (sayakpaul) [inline examples/dreambooth/README_sdxl.md:115]: Do we need to put a disclaimer that training with text encoder is still not possible within a free-tier Colab?

File: examples/dreambooth/train_dreambooth_lora.py [hunk 6]
```diff
@@ -839,6 +839,11 @@ def main(args):
...
+        if args.train_text_encoder:
+            text_encoder.gradient_checkpointing_enable()
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora.py:845]: `enable_gradient_checkpointing()` and `gradient_checkpointing_enable()` will raise eyebrows but no biggie

File: src/diffusers/models/transformer_2d.py (hunk 3 above)
Comment (sayakpaul) [reply src/diffusers/models/transformer_2d.py:295]: Makes sense. 

From the [documentation](https://pytorch.org/docs/stable/checkpoint.html):

> Checkpointing is implemented by rerunning a forward-pass segment for each checkpointed segment during backward. This can cause persistent states like the RNG state to be advanced than they would without checkpointing. By default, checkpointing includes logic to juggle the RNG state such that checkpointed passes making use of RNG (through dropout for example) have deterministic output as compared to non-checkpointed passes. The logic to stash and restore RNG states can incur a moderate performance hit depending on the runtime of checkpointed operations. If deterministic output compared to non-checkpointed passes is not required, supply preserve_rng_state=False to checkpoint or checkpoint_sequential to omit stashing and restoring the RNG state during each checkpoint.

Do we need the comparison part between deterministic output and non-checkpointed output?  If not, we can also disable `preserve_rng_state` no?

File: src/diffusers/pipelines/versatile_diffusion/modeling_text_unet.py [hunk 7]
```diff
@@ -800,7 +800,7 @@ def fn_recursive_set_attention_slice(module: torch.nn.Module, slice_size: List[i
             fn_recursive_set_attention_slice(module, reversed_slice_size)
 
     def _set_gradient_checkpointing(self, module, value=False):
-        if isinstance(module, (CrossAttnDownBlockFlat, DownBlockFlat, CrossAttnUpBlockFlat, UpBlockFlat)):
+        if hasattr(module, "gradient_checkpointing"):
             module.gradient_checkpointing = value
```
Comment (sayakpaul) [inline src/diffusers/pipelines/versatile_diffusion/modeling_text_unet.py:804]: Clean.

**Review by sayakpaul — APPROVED** [review]
Merge away!

Could you maybe do an empty commit to ensure the CI is 🍏?

**Review by williamberman — APPROVED** [review]

**Comment by bghira** [comment]
are you also not interested in fully unloading the text encoders since they aren't actually needed for the training?

**Comment by sayakpaul** [comment]
> are you also not interested in fully unloading the text encoders since they aren't actually needed for the training?

WDYM?

If `train_text_encoder` is not specified, then we always precompute the embeddings and free them up. 

https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_sdxl.py#L930C1-L934C33

File: examples/dreambooth/train_dreambooth_lora_sdxl.py [hunk 8]
```diff
@@ -755,6 +754,12 @@ def main(args):
         else:
             raise ValueError("xformers is not available. Make sure it is installed correctly")
 
+    if args.gradient_checkpointing:
+        unet.enable_gradient_checkpointing()
+        if args.train_text_encoder:
+            text_encoder_one.gradient_checkpointing_enable()
+            text_encoder_two.gradient_checkpointing_enable()
```
Comment (bghira) [inline examples/dreambooth/train_dreambooth_lora_sdxl.py:761]: the pix2pix script uses `text_encoder_1` and _2 which match the actual internal names. can we make them consistent?

File: src/diffusers/models/transformer_2d.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/diffusers/models/transformer_2d.py:295]: think that's ok that's only relevant for customized gradient backprop workflow IMO

File: examples/dreambooth/README_sdxl.md [hunk 9]
```diff
@@ -101,6 +101,24 @@ To better track our training experiments, we're using the following flags in the
...
 
+### Dog toy example with < 16GB VRAM
+
+By making use of [`gradient_checkpointing`](https://pytorch.org/docs/stable/checkpoint.html) (which is natively supported in Diffusers), [`xformers`](https://github.com/facebookresearch/xformers), and [`bitsandbytes`](https://github.com/TimDettmers/bitsandbytes) libraries, you can train SDXL LoRAs with less than 16GB of VRAM by adding the following flags to your accelerate launch command:
+
+```diff
++  --enable_xformers_memory_efficient_attention \
++  --gradient_checkpointing \
++  --use_8bit_adam \
++ --mixed_precision="fp16" \
```
Comment (patrickvonplaten) [inline examples/dreambooth/README_sdxl.md:112]:
```suggestion
+  --mixed_precision="fp16" \
```

**Review by yiyixuxu — APPROVED** [review]
cool!

**Comment by StanislawKarnacky** [comment]
I successfully ran this code with the free version of colab. But processing time of 500 steps in ~3 hours is confusing. I trained a lof of regular Loras, but never trained Dreambooth models, especially within those specific conditions, so, sorry for a noob questions. I took all the settings from the example with dog, but how free am I to change them, without running out of RAM/VRAM? Also are the default/example settings universal? For example i have 77 pictures of my own dog, should i add more or reduce the amount of data? And, as i assume, "--max_train_steps" is number of steps after which the model will be saved? Because usually my training ends with 10 epochs, which gives me the opportunity to test and choose the best one.
Thank you.

**Comment by StanislawKarnacky** [comment]
I give up for now. After ~ 10 attempts, using another VAE, as suggested in the example with a dog, I managed to get a total time of 500 steps for ~ 1h30m on the free version of colab with T4. But every time after all 500 steps I've run out of VRAM or RAM. Sometimes it's 2Mb, sometimes 20. I duplicate my arguments below. If somebody has any suggestions please let me know!
    arguments = [
    "--pretrained_model_name_or_path", "stabilityai/stable-diffusion-xl-base-1.0",
    "--instance_data_dir", "/content/drive/MyDrive/Loras/SDXL/Data",
    "--instance_prompt", "A photo of sks dog",,
    "--output_dir", "/content/drive/MyDrive/Loras/SDXL/Output",
    "--seed", "0",
    "--train_batch_size", "1",
    "--learning_rate", "1e-4",
    "--lr_scheduler", "constant",
    "--mixed_precision", "fp16",
    "--resolution", "1024",  # Corrected the resolution argument with a comma (,)
    "--gradient_accumulation_steps", "4",  # Converted integer value to string
    "--lr_warmup_steps", "0",  # Converted integer value to string
    "--max_train_steps", "500",  # Converted integer value to string
    "--validation_prompt", "A photo of sks dog",
    "--validation_epochs", "25",  # Converted integer value to string
    "--enable_xformers_memory_efficient_attention",
    "--gradient_checkpointing",
    "--use_8bit_adam",
    "--pretrained_vae_model_name_or_path", "madebyollin/sdxl-vae-fp16-fix"
    ]

**Comment by sayakpaul** [comment]
Does this help?

https://colab.research.google.com/drive/1CvDFD-vDohrerhpiNRLtn32ON2ARgBnX?usp=sharing

**Comment by StanislawKarnacky** [comment]
> Does this help?
> 
> https://colab.research.google.com/drive/1CvDFD-vDohrerhpiNRLtn32ON2ARgBnX?usp=sharing

Yes, it is, thank you! I don't know how did you manage to do that. In my case, I just used code from "train_dreambooth_lora_sdxl.py" and used suggested args and "from accelerate.utils import write_basic_config".
Btw it there any chance to convert .bin to .safetensors? Tried a couple of scripts, keys mismatch.

**Comment by dai-ichiro** [comment]
> I give up for now. After ~ 10 attempts, using another VAE, as suggested in the example with a dog, I managed to get a total time of 500 steps for ~ 1h30m on the Google Colab with T4. But every time after all 500 steps I've run out of VRAM or RAM. Sometimes it's 2Mb, sometimes 20. I duplicate my arguments below. If somebody has any suggestions please let me know! arguments = [ "--pretrained_model_name_or_path", "stabilityai/stable-diffusion-xl-base-1.0", "--instance_data_dir", "/content/drive/MyDrive/Loras/SDXL/Data", "--instance_prompt", "A photo of sks dog",, "--output_dir", "/content/drive/MyDrive/Loras/SDXL/Output", "--seed", "0", "--train_batch_size", "1", "--learning_rate", "1e-4", "--lr_scheduler", "constant", "--mixed_precision", "fp16", "--resolution", "1024", # Corrected the resolution argument with a comma (,) "--gradient_accumulation_steps", "4", # Converted integer value to string "--lr_warmup_steps", "0", # Converted integer value to string "--max_train_steps", "500", # Converted integer value to string "--validation_prompt", "A photo of sks dog", "--validation_epochs", "25", # Converted integer value to string "--enable_xformers_memory_efficient_attention", "--gradient_checkpointing", "--use_8bit_adam", "--pretrained_vae_model_name_or_path", "madebyollin/sdxl-vae-fp16-fix" ]

I have same error.
Default value of checkpointing_steps arg is 500. That may be  cause of your OOM.
In my case, when training reach first checkpointint_steps, OOM occurs.

**Comment by sayakpaul** [comment]
Try using a larger checkpointing steps than your total maximum training steps.

**Comment by dai-ichiro** [comment]
Thank you for your quick response.

When I use a larger checkpointing steps than total maximum training steps, training works fine.

I want to resume training from the saved checkpoints, but I don't know how to.
Is there a way to resume training without using checkpointing_steps?

**Comment by sayakpaul** [comment]
* Not sure how that would be possible.
* For checkpointing, use a larger GPU if possible.

**Comment by dai-ichiro** [comment]
Thank you very much.
I will try it soon.

**Comment by bghira** [comment]
checkpointing likely consumes a lot of VRAM because image validations end up loading the text encoder again.
